scalyr_agent/tests/copying_manager_test.py

```diff
@@ -784,6 +821,186 @@ def test_pipelined_requests_with_retry_error(self):
 
         responder_callback('success')
 
+    def test_start_from_full_checkpoint(self):
+        controller = self.__create_test_instance()
+        previous_root_dir = os.path.dirname(self.__test_log_file)
+
+        self.__append_log_lines('First line', 'Second line')
+        (request, responder_callback) = controller.wait_for_rpc()
+        lines = self.__extract_lines(request)
+
+        self.assertEquals(2, len(lines))
+        self.assertEquals('First line', lines[0])
+        self.assertEquals('Second line', lines[1])
+
+        # stop thread on manager to write checkouts to file.
+        controller.stop()
+
+        # write some new lines to log.
+        self.__append_log_lines('Third line', 'Fourth line')
+
+        # create new copying manager, but passing previous directory with same log and checkouts.
+        controller = self.__create_test_instance(root_dir=previous_root_dir, auto_start=False)
+
+        self._manager.start_manager(dict(fake_client=True))
```

Comment:
Why do you need the `auto_start=False` but then immediately start it here?

Reply:
The problem was that when copying manager starts, it stops only at "SLEEPING" state. https://github.com/scalyr/scalyr-agent-2/blob/8726d6c2483c9ceb19a37697f266261c31e00b0b/scalyr_agent/tests/copying_manager_test.py#L1144-L1157. This means, it has to do full iteration before we can stop on needed state, and we can't intercept the first "SENDING".
The solution was to not allow it to do its first iteration, by disabling auto start in constructor and by starting it manually. After, I decided to pull automated start from controller's constructor to test cases. I think it is clearer and there is no need to make additional logic to constructors.